### PR TITLE
Add deprecations to new parser

### LINKED ISSUE #5119: Move deprecation warnings into the new parser
**Issue**

We are parsing a config file twice, once when creating the ert_config object and again when trying to generate the list of suggestions. This should be refactored to only do the parsing once.

### PR DESCRIPTION
**Pull request by yngve-sk**
**Issue**
Resolves #5119 

**Approach**
_Short description of the approach_

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [ ] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: tests/test_config_parsing/test_parser_error_collection.py [hunk 1]
```diff
@@ -160,25 +161,26 @@ def assert_that_config_leads_to_warning(
     with pytest.warns(ConfigWarning) as the_warnings:
         ErtConfig.from_file(config_filename, use_new_parser=True)
 
-    warning_infos: List[WarningInfo] = []
+    warning_infos: List[ErrorInfo] = []
     for x in the_warnings.list:
-        assert isinstance(x, Warning)
-        warning_info: WarningInfo = x.args[0]
+        assert isinstance(x, WarningMessage)
+        message = cast(ConfigWarning, x.message)
```
Comment (andreas-el) [inline tests/test_config_parsing/test_parser_error_collection.py:167]: Is this to silence mypy? I fail to see what it does here, as you could have done `x.message.info` instead?

File: src/ert/parsing/ext_job_schema.py [hunk 2]
```diff
@@ -180,14 +182,38 @@ def portable_exe_keyword() -> SchemaItem:
     arg_type_keyword(),
     env_keyword(),
     exec_env_keyword(),
-    portable_exe_keyword(),
+    # portable_exe_keyword(),
```
Comment (andreas-el) [inline src/ert/parsing/ext_job_schema.py:185]: Remove this

File: src/ert/parsing/config_schema_deprecations.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        "was removed in 2015, and this keyword can be safely removed from "
+        "the config file.",
+    ),
+    DeprecationInfo(
+        keyword="RFTPATH",
+        message="The RFTPATH keyword was used to give a path to well observations "
+        "to be used for plotting. The corresponding plotting functionality "
+        "was removed in 2015, and the RFTPATH keyword can safely be removed "
+        "from the config file.",
+    ),
```
Comment (andreas-el) [inline src/ert/parsing/config_schema_deprecations.py:103]: We plan to remove the already existing suggestions too?

File: src/ert/parsing/config_schema_deprecations.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    "MULTZ",
+    "EQUIL",
+    "GEN_PARAM",
+    "MULTFLT",
+]
+JUST_REMOVE_KEYWORDS = [
+    "UMASK",
+    "LOG_FILE",
+    "LOG_LEVEL",
+    "ENKF_RERUN",
```
Comment (andreas-el) [inline src/ert/parsing/config_schema_deprecations.py:14]: You omit `GEN_KW_TAG_FORMAT`, is that on purpose?

File: src/ert/parsing/config_schema_deprecations.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            "simply remove this from your config. If you need to set these "
+            "options it is now done via the QUEUE_OPTION keyword.",
+        )
+        for kw in USE_QUEUE_OPTION
+    ],
+    DeprecationInfo(
+        keyword="SCHEDULE_PREDICTION_FILE",
+        message="The 'SCHEDULE_PREDICTION_FILE' config keyword "
+        "has been removed and no longer has any effect",
+    ),
```
Comment (andreas-el) [inline src/ert/parsing/config_schema_deprecations.py:83]: Is this a duplicate? See at the bottom of this list

File: src/ert/parsing/schema_dict.py [hunk 6]
```diff
@@ -51,5 +39,70 @@ def check_required(
...
+
+        def push_deprecation(info: DeprecationInfo, line: List[ContextString]):
+            if info.check is None or (callable(info.check) and info.check(line)):
+                detected_deprecations.append((info, line))
+
+        for kw, v in config_dict.items():
+            deprecation_info = maybe_deprecated_kws_dict.get(kw)
+            if deprecation_info and kw in self:
+                if v is None:
+                    # Edge case:
```
Comment (andreas-el) [inline src/ert/parsing/schema_dict.py:74]: Do you have an example for an edge case that would fail?

**Review by andreas-el — CHANGES_REQUESTED** [review]
Looks good, but needs some small tweaking

File: src/ert/parsing/config_schema_deprecations.py (hunk 4 above)
Comment (yngve-sk) [reply src/ert/parsing/config_schema_deprecations.py:14]: Oops!! Not on purpose, I put it back

File: src/ert/parsing/config_schema_deprecations.py (hunk 5 above)
Comment (yngve-sk) [reply src/ert/parsing/config_schema_deprecations.py:83]: Yep, removing it

File: src/ert/parsing/ext_job_schema.py (hunk 2 above)
Comment (yngve-sk) [reply src/ert/parsing/ext_job_schema.py:185]: Done

File: tests/test_config_parsing/test_parser_error_collection.py [hunk 7]
```diff
@@ -901,3 +925,146 @@ def test_that_executable_directory_errors(dirname):
...
+SCHEDULE_PREDICTION_FILE
+HAVANA_FAULT
+REFCASE_LIST
+RFTPATH
+END_DATE
+CASE_TABLE
+RERUN_START
+DELETE_RUNPATH
+PLOT_SETTINGS
+UPDATE_PATH
```
Comment (andreas-el) [inline tests/test_config_parsing/test_parser_error_collection.py:967]: I think you break your own test when removing the arguments, UPDATE_PATH require two arguments.

**Review by andreas-el — CHANGES_REQUESTED** [review]
Need to mixup tests by resolving UPDATE_PATH issue.

File: src/ert/parsing/config_schema_deprecations.py (hunk 3 above)
Comment (andreas-el) [reply src/ert/parsing/config_schema_deprecations.py:104]: I've created a note for this in https://github.com/equinor/ert/issues/5355

File: tests/test_config_parsing/test_parser_error_collection.py [hunk 8]
```diff
@@ -158,15 +158,10 @@ def assert_that_config_leads_to_warning(
         {config_filename: config_file_contents, **(expected_error.other_files or {})}
     )
 
-    with pytest.warns(ConfigWarning) as the_warnings:
-        ErtConfig.from_file(config_filename, use_new_parser=True)
+    # with pytest.warns(ConfigWarning) as the_warnings:
```
Comment (andreas-el) [inline tests/test_config_parsing/test_parser_error_collection.py:161]: Remove?

File: src/ert/parsing/lark_parser.py [hunk 9]
```diff
@@ -529,10 +529,19 @@ def parse(
...
+            pre_defines=pre_defines,
+            tree=tree,
+            site_config=site_config,
+            schema=schema,
+        )
+
+        config_warnings: List[ConfigWarning] = [w.message for w in warnings_list]
+        warning_infos = [w.info for w in config_warnings]
+
+        config_dict.warning_infos = warning_infos
```
Comment (andreas-el) [inline src/ert/parsing/lark_parser.py:545]: Combine these?

File: tests/test_config_parsing/test_parser_error_collection.py [hunk 10]
```diff
@@ -158,15 +158,10 @@ def assert_that_config_leads_to_warning(
...
+    ert_config = ErtConfig.from_file(config_filename, use_new_parser=True)
 
-    warning_infos: List[ErrorInfo] = []
-    for x in the_warnings.list:
-        assert isinstance(x, WarningMessage)
-        message = cast(ConfigWarning, x.message)
-        warning_info = message.info
-        warning_infos.append(warning_info)
+    warning_infos = ert_config.warning_infos
```
Comment (andreas-el) [inline tests/test_config_parsing/test_parser_error_collection.py]: Combine?

File: src/ert/parsing/config_schema_deprecations.py [hunk 11]
```diff
@@ -25,10 +26,17 @@
     *[
         DeprecationInfo(
             keyword=kw,
-            message=f"Using {kw} with substitution strings"
+            message=lambda line: f"Using {kw} with substitution strings"
```
Comment (andreas-el) [inline src/ert/parsing/config_schema_deprecations.py:29]: Verify this lambda, lint provides this warning;
`Function definition does not bind loop variable 'kw'.`

File: src/ert/parsing/types.py [hunk 12]
```diff
@@ -19,6 +19,5 @@
 MaybeWithContext = Union[ContextValue, Primitives, FileContextToken]
 
 
-ConfigDict = Dict[
-    str, Union[MaybeWithContext, List[MaybeWithContext], List[List[MaybeWithContext]]]
-]
+class ConfigDict(Dict[str, Union[str, List[str], List[List[str]]]]):
+    warning_infos: Optional[List["WarningInfo"]] = None
```
Comment (andreas-el) [inline src/ert/parsing/types.py:23]: Need to include this? Lint errors out on this one.

**Review by andreas-el — CHANGES_REQUESTED** [review]
Still more things to resolve, -- getting better!

File: tests/test_config_parsing/test_parser_error_collection.py (hunk 8 above)
Comment (yngve-sk) [reply tests/test_config_parsing/test_parser_error_collection.py:160]: Done

File: src/ert/parsing/lark_parser.py (hunk 9 above)
Comment (yngve-sk) [reply src/ert/parsing/lark_parser.py:545]: Done

File: tests/test_config_parsing/test_parser_error_collection.py (hunk 10 above)
Comment (yngve-sk) [reply tests/test_config_parsing/test_parser_error_collection.py]: Done(?)

File: src/ert/parsing/config_schema_deprecations.py (hunk 11 above)
Comment (yngve-sk) [reply src/ert/parsing/config_schema_deprecations.py]: Fixed

**Pull request merged**